### PR TITLE
AOTDispatcher: limit cases when we detach() graph inputs to non-leaves

### PR DESCRIPTION
**Pull request by bdhirsh**
This PR is slightly a revival / update to the discussion from https://github.com/pytorch/pytorch/pull/98960:

Part of FSDP2's tracing strategy right now is that:

(1) it is painful/difficult to handle the case where we have multiple graph input tensors that are aliased to each other and at least one of them is duplicated

(2) we already have longstanding in logic to remove duplicate input tensors from the graph in dynamo. Morally, FSDP2 gives us duplicate input tensors in the backward graph for every `unsharded_param`, because we have (a) the `unsharded_param` being closed over by the backward hook to resize/allgather, and (b) the same `unsharded_param` being saved for backward by autograd (we now guarantee in the partitioner that we will always save the base tensor for backward and recompute views)

(3) However, we were still seeing cases where the `unsharded_param` showed up twice in the backward graph inputs, as distinct tensor objects (with different python ids) instead of being true duplicates that dynamo can de-dup.

It turns on that this was because we were `.detach()`ing the `unsharded_param` in AOTDispatcher before plumbing it through the compiled forward (and so autograd would save a detach'd version of the `unsharded_param`). This is precisely because of the logic from https://github.com/pytorch/pytorch/pull/98960.

However, re-reading the detailed comments, it seems unnecessary to do a detach() on a graph input that is a (leaf) `nn.Parameter`, even if it happens to get no gradients in the backward. Since it is a leaf, we don't have to worry about the autograd engine "continuing to backprop through the graph beyond the current tensor" (the leaf has no other grad_fn for autograd to backprop through).

So this PR makes us a bit less aggressive about calling detach() on inputs: we only do it when:

(1) our graph input statically will get a `None` gradient (and also has no metadata mutations, the existing state)

(2) **and** our graph input is a non-leaf tensor (so detach()ing is actually required to prevent autograd from incorrectly backpropping past the non-leaf.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #134997
* #131507
* #133730
* __->__ #134193

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/testing/_internal/optests/aot_autograd.py [hunk 1]
```diff
@@ -118,7 +118,8 @@ def check(args, ignore_failure=False):
             raise
 
         # See https://github.com/pytorch/pytorch/pull/98960#issuecomment-1505962215
-        if all(x is None for x in orig_grad):
+        any_non_leaves = any(x.requires_grad and not x.is_leaf for x in args)
+        if all(x is None for x in orig_grad) and any_non_leaves:
```
Comment (bdhirsh) [inline torch/testing/_internal/optests/aot_autograd.py:122]: cc @albanD, I'm hoping what I put in the PR description makes sense to you :) (after re-reading some of the previous commentary)

File: torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py [hunk 2]
```diff
@@ -501,7 +501,11 @@ def aot_dispatch_autograd(
                     == MutationType.MUTATED_IN_GRAPH
                     and fw_metadata.input_info[i].mutates_storage_metadata
                 )
-                if bw_out is None and not metadata_mutation_in_graph:
+                is_non_leaf = (
```
Comment (albanD) [inline torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py:504]: Another way to look at this one is `is_inside_autograd_graph/has_autograd_history = input.grad_fn is not None` (if you have that info here).
I would rename the variable either way as, as the condition shows, you are no using the standard "leaf" definition here.

File: torch/testing/_internal/optests/aot_autograd.py [hunk 3]
```diff
@@ -118,7 +118,9 @@ def check(args, ignore_failure=False):
             raise
 
         # See https://github.com/pytorch/pytorch/pull/98960#issuecomment-1505962215
-        if all(x is None for x in orig_grad):
+        tensor_args = [x for x in pytree.tree_flatten(args)[0] if isinstance(x, torch.Tensor)]
+        any_non_leaves = any(x.requires_grad and not x.is_leaf for x in tensor_args)
```
Comment (albanD) [inline torch/testing/_internal/optests/aot_autograd.py:122]: Same as above:
```suggestion
        any_non_leaves = any(x.grad_fn is not Nonefor x in tensor_args)
```

**Review by albanD — COMMENTED** [review]
Makes sense!

File: torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py (hunk 2 above)
Comment (bdhirsh) [reply torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py:504]: yes, agreed with the name change

File: torch/testing/_internal/optests/aot_autograd.py (hunk 3 above)
Comment (yf225) [reply torch/testing/_internal/optests/aot_autograd.py:122]: also an interesting error from CI:
```
2024-08-22T07:36:12.9280359Z [31m[1m_ TestEagerFusionModuleInfoCPU.test_aot_autograd_module_exhaustive_nn_AdaptiveAvgPool1d_cpu_float32 _[0m
2024-08-22T07:36:12.9281197Z Traceback (most recent call last):
2024-08-22T07:36:12.9282087Z   File "/var/lib/jenkins/workspace/test/functorch/test_aotdispatch.py", line 6251, in test_aot_autograd_module_exhaustive
2024-08-22T07:36:12.9283196Z     _test_aot_autograd_module_helper(self, device, dtype, training, module_info)
2024-08-22T07:36:12.9284295Z   File "/var/lib/jenkins/workspace/test/functorch/test_aotdispatch.py", line 6184, in _test_aot_autograd_module_helper
2024-08-22T07:36:12.9285234Z     _test_aot_autograd_forwards_backwards_helper(
2024-08-22T07:36:12.9286567Z   File "/opt/conda/envs/py_3.12/lib/python3.12/site-packages/torch/testing/_internal/optests/aot_autograd.py", line 139, in _test_aot_autograd_forwards_backwards_helper
2024-08-22T07:36:12.9287803Z     check(args, ignore_failure=False)
2024-08-22T07:36:12.9288843Z   File "/opt/conda/envs/py_3.12/lib/python3.12/site-packages/torch/testing/_internal/optests/aot_autograd.py", line 121, in check
2024-08-22T07:36:12.9289938Z     any_non_leaves = any(x.requires_grad and not x.is_leaf for x in args)
2024-08-22T07:36:12.9290553Z                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-08-22T07:36:12.9291657Z   File "/opt/conda/envs/py_3.12/lib/python3.12/site-packages/torch/testing/_internal/optests/aot_autograd.py", line 121, in <genexpr>
2024-08-22T07:36:12.9292765Z     any_non_leaves = any(x.requires_grad and not x.is_leaf for x in args)
2024-08-22T07:36:12.9293338Z                          ^^^^^^^^^^^^^^^
2024-08-22T07:36:12.9293931Z AttributeError: 'dict' object has no attribute 'requires_grad'
2024-08-22T07:36:12.9294347Z 
2024-08-22T07:36:12.9294617Z To execute this test, run the following from the base repo dir:
2024-08-22T07:36:12.9295763Z     python test/functorch/test_aotdispatch.py TestEagerFusionModuleInfoCPU.test_aot_autograd_module_exhaustive_nn_AdaptiveAvgPool1d_cpu_float32
```

Comment (bdhirsh) [reply torch/testing/_internal/optests/aot_autograd.py:122]: hmm there are some new CI errors  I need to look into, but I think I fixed that one (update the PR to pytree_flatten and only look at tensor args)

**Comment by bdhirsh** [comment]
Some of the failures I can't repro locally (I'll do another sanity run in CI), but the main test failing for me is:
```
python test/inductor/test_distributed_patterns.py DistributedPatternTests.test_fake_distributed_inductor
```

Looking at it and diffing the graphs produced by compiled autograd with/without my PR, I see:
```
INFO: TRACED GRAPH
 ===== Compiled autograd graph =====
 <eval_with_key>.3 class CompiledAutograd(torch.nn.Module):
    def forward(self, inputs, sizes, scalars, hooks):
        # No stacktrace found for following nodes
        getitem: "f32[][]cuda:0" = inputs[0]
        aot0_primals_1: "f32[10, 20][20, 1]cuda:0" = inputs[1]
        aot0_primals_4: "f32[2, 20][20, 1]cuda:0" = inputs[2]
        getitem_3: "f32[s3, 20][20, 1]cuda:0" = inputs[3];  inputs = getitem_3 = None

         # File: /home/hirsheybar/local/a/pytorch/torch/_dynamo/compiled_autograd.py:479 in set_node_origin, code: SumBackward0 (NodeCall 1)
        expand: "f32[2, 10][0, 0]cuda:0" = torch.ops.aten.expand.default(getitem, [2, 10]);  getitem = None
```

Where the `getitem_3` is an extra node that my PR produces for some reason, that is unused in the graph. Still digging into why my PR causes compiled autograd to create that extra node.

**Comment by bdhirsh** [comment]
A few updates:

Just for clarity, the failing test:
`python test/inductor/test_distributed_patterns.py DistributedPatternTests.test_fake_distributed_inductor`

gives this error:
```
  File "/home/hirsheybar/local/a/pytorch/torch/autograd/graph.py", line 817, in _engine_run_backward
    return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
  File "/home/hirsheybar/local/a/pytorch/torch/_dynamo/compiled_autograd.py", line 337, in runtime_wrapper
    return compiled_fn(inputs, sizes, scalars, hooks)
  File "/home/hirsheybar/local/a/pytorch/torch/nn/modules/module.py", line 1736, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/hirsheybar/local/a/pytorch/torch/nn/modules/module.py", line 1747, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/hirsheybar/local/a/pytorch/torch/_dynamo/eval_frame.py", line 469, in _fn
    return fn(*args, **kwargs)
  File "/home/hirsheybar/local/a/pytorch/torch/fx/graph_module.py", line 785, in call_wrapped
    return self._wrapped_call(self, *args, **kwargs)
  File "/home/hirsheybar/local/a/pytorch/torch/fx/graph_module.py", line 360, in __call__
    raise e.with_traceback(None)  # noqa: B904
IndexError: list index out of range
```

I'm still trying to understand what's going on (@xmfan in case you have any ideas?). But one very suspicious thing that I found is that deleting all of the list clearing callsites "fixes" this problem. I wonder if somehow, after we run the backward the first time, we clear the list and dynamo recompiles the compiled autograd graph but with an empty arg list, causing the out of range.

**Comment by bdhirsh** [comment]
(in particular, I get the test passing with this 2-liner):
```
diff --git a/torch/_dynamo/utils.py b/torch/_dynamo/utils.py
index 27a45076646..be44e450d8d 100644
--- a/torch/_dynamo/utils.py
+++ b/torch/_dynamo/utils.py
@@ -3045,8 +3045,8 @@ def flatten_graph_inputs(gm: torch.fx.GraphModule, inputs, compile_gm):
         flat_args = flatten_fn(args)

         # flat_args is a new list, so we need to clear references from the old list
-        for i in inputs_idx_to_clear:
-            args[i].clear()
+        #for i in inputs_idx_to_clear:
+            #args[i].clear()
```

**Comment by xmfan** [comment]
> Where the getitem_3 is an extra node that my PR produces for some reason, that is unused in the graph

This looks strange, looking at the dynamo graph, there's some `torch.ops.fsdp.set_` op that uses it, but the op was inserted during dynamo tracing: https://interncache-all.fbcdn.net/manifold/tlparse_reports/tree/logs/.tmpDjvDBT/index.html

The test also passes the aot_eager variant, but when I put breakpoints it fails for different reasons, incl. CUDA IMA

**Comment by bdhirsh** [comment]
> this looks strange, looking at the dynamo graph, there's some torch.ops.fsdp.set_ op that uses it, but the op was inserted during dynamo tracing: https://interncache-all.fbcdn.net/manifold/tlparse_reports/tree/logs/.tmpDjvDBT/index.html

my guess is that this is:  `getitem_3` is a model param, that is being secretly-closed-over in a backward hook (in a way that you can't tell from the compiled-autograd graph, but you can see once dynamo runs and desugars the backward hook

> The test also passes the aot_eager variant, but when I put breakpoints it fails for different reasons, incl. CUDA IMA

🤔 i was doing quite a few weird things locally including lots of breakpoints and managed to not see an IMA

**Comment by yf225** [comment]
@bdhirsh @xmfan I fixed the `test_fake_distributed_inductor` to be more representative of the actual FSDP2 program. I believe it removes the recompile and the test passes now 🙂

**Review by yf225 — APPROVED** [review]

**Comment by bdhirsh** [comment]
CI still failing for some inductor opinfos, looking

**Comment by yf225** [comment]
@bdhirsh we might also need to patch https://gist.github.com/yf225/7d6b273dcb85a9963faadc2aa6177016 to fix the `test_fake_distributed_inductor` error

**Pull request closed without merging**